dwave/preprocessing/src/presolveimpl.hpp

```diff
@@ -84,12 +131,52 @@ class PresolverImpl {
         changes |= normalization_remove_self_loops();
         changes |= normalization_flip_constraints();
         changes |= normalization_remove_invalid_markers();
+        changes |= normalization_fix_bounds();
 
         normalized_ = true;
 
         return changes;
     }
 
+    /// Normalize the variable bounds by vartype and check that lb <= ub
+    bool normalization_fix_bounds() {
+        bool changes = false;
+        for (size_type v = 0; v < model_.num_variables(); ++v) {
+            // tighten bounds based on the vartype
+            switch (model_.vartype(v)) {
+                case dimod::Vartype::SPIN:
+                    throw std::logic_error(
+                            "normalization_fix_bounds() must be run after "
+                            "normalization_spin_to_binary()");
+                case dimod::Vartype::BINARY:
+                    changes |= model_.set_upper_bound(v, 1);
+                    changes |= model_.set_lower_bound(v, 0);
+
+                    // we carry on into INTEGER to handle fractional
+
+                case dimod::Vartype::INTEGER:
+                    if (std::ceil(model_.lower_bound(v)) > std::floor(model_.upper_bound(v))) {
+                        throw InvalidModelError(
+                                "variable lower bound must be less than or equal to upper bound");
+                    }
+
+                    changes |= model_.set_upper_bound(v, std::floor(model_.upper_bound(v)));
+                    changes |= model_.set_lower_bound(v, std::ceil(model_.lower_bound(v)));
+                    break;
+                case dimod::Vartype::REAL:
+                    break;
+            }
+
```

Comment:
I believe that switches should (almost always) have a default clause, even if it just throws an exception.  I've seen too many hard to track-down bugs because a case is added that wasn't handled

Reply:
We're switching over an enum, so the compiler will complain if there is a missing case, unless we had a default in which case the missing case would pass silently. IMO this way it's more clear when something is missing than adding the default.